Lib/importlib/_bootstrap.py

```diff
@@ -1020,17 +1020,26 @@ def _find_and_load_unlocked(name, import_):
 
 def _find_and_load(name, import_):
     """Find and load the module."""
-    with _ModuleLockManager(name):
-        module = sys.modules.get(name, _NEEDS_LOADING)
-        if module is _NEEDS_LOADING:
-            return _find_and_load_unlocked(name, import_)
+    module = sys.modules.get(name, _NEEDS_LOADING)
+    if (
+        module is _NEEDS_LOADING
+        or (
+            hasattr(module, "__spec__")
+            and getattr(module.__spec__, "_initializing", False)
+        )
+    ):
+        with _ModuleLockManager(name):
+            module = sys.modules.get(name, _NEEDS_LOADING)
+            if module is _NEEDS_LOADING:
+                return _find_and_load_unlocked(name, import_)
+
+        _lock_unlock_module(name)
```

Comment:
This optimization looks a lot like what is done in `import_ensure_initialized` in `Python/import.c`, can I ask you to add similar comments to those in that function?

Reply:
Done! I added a few comments about the optimizations and tweaked the code as suggested.